pkg/reconciler/identity/iam_policy.go

```diff
@@ -0,0 +1,238 @@
+	Remove
+)
+
+// GServiceAccount identifies a Google service account by project and name.
+type GServiceAccount struct {
+	ProjectID string
+	Name      string
+}
+
+type modificationRequest struct {
+	serviceAccount GServiceAccount
+	role           iam.RoleName
+	member         string
+	action         action
+	respCh         chan error
+}
+
+type roleModification struct {
+	addMembers    map[string]bool
+	removeMembers map[string]bool
+}
+
+type batchedModifications struct {
+	roleModifications map[iam.RoleName]*roleModification
+	listeners         []chan<- error
+}
+
+type getPolicyResponse struct {
+	account GServiceAccount
+	policy  *iam.Policy
+	err     error
+}
+
+type setPolicyResponse struct {
+}
+
+// IAMPolicyManager serializes and batches IAM policy changes to a Google
+// Service Account to avoid conflicting changes.
+type IAMPolicyManager struct {
+	iam         gclient.IamClient
+	requestCh   chan *modificationRequest
+	pending     map[GServiceAccount]*batchedModifications
+	getPolicyCh chan *getPolicyResponse
+}
+
+// NewIAMPolicyManager creates an IAMPolicyManager using the given
+// IamClient. The IAMPolicyManager will execute until ctx is cancelled.
+func NewIAMPolicyManager(ctx context.Context, client gclient.IamClient) (*IAMPolicyManager, error) {
+	m := &IAMPolicyManager{
+		iam:         client,
+		requestCh:   make(chan *modificationRequest),
+		pending:     make(map[GServiceAccount]*batchedModifications),
+		getPolicyCh: make(chan *getPolicyResponse),
+	}
+	go m.manage(ctx)
+	return m, nil
+}
+
+// AddIAMPolicyBinding adds or updates an IAM policy binding for the given account
+// and role to include member. This call will block until the IAM update
+// succeeds or fails or until ctx is cancelled.
+func (m *IAMPolicyManager) AddIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role iam.RoleName) error {
+	return m.doRequest(ctx, &modificationRequest{
+		serviceAccount: account,
+		role:           role,
+		member:         member,
+		action:         Add,
+		respCh:         make(chan error, 1),
+	})
+}
+
+// RemoveIAMPolicyBinding removes or updates an IAM policy binding for the given
+// account and role to remove member. This call will block until the IAM update
+// succeeds or fails or until ctx is cancelled.
+func (m *IAMPolicyManager) RemoveIAMPolicyBinding(ctx context.Context, account GServiceAccount, member string, role iam.RoleName) error {
+	return m.doRequest(ctx, &modificationRequest{
+		serviceAccount: account,
+		role:           role,
+		member:         member,
+		action:         Remove,
+		respCh:         make(chan error, 1),
+	})
+}
+
+func (m *IAMPolicyManager) doRequest(ctx context.Context, req *modificationRequest) error {
+	select {
+	case m.requestCh <- req:
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+	select {
+	case err := <-req.respCh:
+		return err
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (m *IAMPolicyManager) manage(ctx context.Context) {
+	for {
+		select {
```

Comment:
Please add some documentation about how this is expected to work. E.g. `AddIAMPolicyBinding` is called, which puts something into the `requestCh`, which ...

Reply:
Done.